EMode (v1)

0xLendlord opened the pull request:
- reserve config module
- lending_market::set_emode_status
- obligation::refresh

0xLendlord committed:
- reserve config module
- lending_market::set_emode_status
- obligation::refresh

0xripleys commented on contracts/suilend/sources/reserve_config.move:
```diff
@@ -57,6 +64,13 @@ module suilend::reserve_config {
         fields: Bag
     }
 
+    public struct EModeData has store, copy, drop {
```
nit: EModeEntry

0xripleys commented on contracts/suilend/sources/reserve_config.move:
```diff
@@ -156,6 +237,21 @@ module suilend::reserve_config {
+            while (vector::length(&keys) > 0) {
+                let emode = vec_map::get(emode_ltvs, &vector::pop_back(&mut keys));
+
+                assert!(config.open_ltv_pct < emode.open_ltv_pct, ENormalOpenLtvBetterThanEModeLtvs);
```
<=.

0xripleys commented on contracts/suilend/sources/reserve_config.move:
```diff
@@ -293,6 +398,10 @@ module suilend::reserve_config {
         set_open_attributed_borrow_limit_usd(&mut builder, config.open_attributed_borrow_limit_usd);
         set_close_attributed_borrow_limit_usd(&mut builder, config.close_attributed_borrow_limit_usd);
 
+        if (bag::contains(&config.additional_fields, EModeKey {})) {
```
this doesn't work. can you test?

0xLendlord replied:
done

0xripleys commented on contracts/suilend/sources/obligation.move:
```diff
@@ -157,6 +169,30 @@ module suilend::obligation {
+        assert!(vector::length(&obligation.borrows) <= 1, EEModeNotValidWithCrossMargin);
+        assert!(vector::length(&obligation.deposits) <= 1, EEModeNotValidWithCrossMargin);
+
+        if (vector::length(&obligation.deposits) == 1) {
```
don't need this branch, but keep the refresh

0xripleys replied:
also need a way to toggle emode

0xripleys commented on contracts/suilend/sources/obligation.move:
```diff
@@ -169,6 +205,7 @@ module suilend::obligation {
         let mut deposited_value_usd = decimal::from(0);
         let mut allowed_borrow_value_usd = decimal::from(0);
         let mut unhealthy_borrow_value_usd = decimal::from(0);
+        let is_emode = is_emode(obligation);
```
lets just make a diff function called obligation::refresh_emode()

0xripleys commented on contracts/suilend/sources/reserve.move:
```diff
@@ -1031,4 +1031,29 @@ module suilend::reserve {
+        close_ltv_pct: u8,
+        ctx: &mut TxContext,
+    ) {
+        let config = cell::get_mut(&mut reserve.config);
```
refactor to not use cell:get_mut, then delete cell::get_mut

0xLendlord committed:
Review items

0xLendlord committed:
Review items

0xripleys commented on contracts/suilend/sources/obligation.move:
```diff
@@ -178,18 +178,7 @@ module suilend::obligation {
-        };
+        refresh(obligation, reserves, clock);
 
         df::add(&mut obligation.id, EModeFlag {}, true);
```
you need to refresh after adding the flag for this to be correct

0xripleys commented on contracts/suilend/sources/lending_market.move:
```diff
@@ -378,6 +378,26 @@ module suilend::lending_market {
+    public fun set_emode<P>(
+        lending_market: &mut LendingMarket<P>,
+        obligation_owner_cap: &ObligationOwnerCap<P>,
+        clock: &Clock
```
ability to toggle emode is still missing

0xripleys commented on contracts/suilend/sources/obligation.move:
```diff
@@ -169,7 +194,11 @@ module suilend::obligation {
         let mut unhealthy_borrow_value_usd = decimal::from(0);
+        let is_emode = is_emode(obligation);
+        if (is_emode) {
+            obligation.refresh_emode(reserves, is_emode, clock)
```
nit: just return immediately

0xLendlord replied:
assigned it to a variable to avoid borrow-checker error

0xripleys replied:
```rust
if (is_emode) {
        obligation.refresh_emode(reserves, is_emode, clock)
        return;
}
```

0xripleys commented on contracts/suilend/sources/obligation.move:
```diff
@@ -260,6 +289,126 @@ module suilend::obligation {
+        };
+    }
+    
+    fun refresh_emode<P>(
```
i think you can simplify this function to not use a while loops at all, and the variables on 303-305 aren't necessary either because the while loop will only iterate at most 1 time

0xLendlord replied:
Simplified

0xripleys commented on contracts/suilend/sources/obligation.move:
```diff
@@ -260,6 +289,126 @@ module suilend::obligation {
+            borrow.market_value = market_value;
+            unweighted_borrowed_value_usd = add(unweighted_borrowed_value_usd, market_value);
+
+            let borrow_weight = if (is_emode) {
```
i think even if obligation is emode, if there's no corresponding emode entry, then the borrow weight has to be the one stored on the reserve (ie not 1). 

example: say user is depositing usdc and borrowing FUD. fud has a borrow weight of 2. the (usdc, fud) pair does not have a corresponding emode entry. then, if the user sets their obligation to emode, then the borrow weight being used is 1, which lets them borrow more fud than we intended.

0xLendlord replied:
Good point. Refactored

0xripleys commented on contracts/suilend/sources/obligation.move:
```diff
@@ -1221,4 +1370,57 @@ module suilend::obligation {
+            // check_normal_ltv_against_emode_ltvs
+            let borrow_reserve_index = get_single_borrow_array_reserve_if_any(obligation);
+
+            assert!(
```
we never want this assert to get tripped because then obligation::refresh can abort, which would be very bad.

0xripleys replied:
in this case we still want to return the base ltvs, right?

0xLendlord replied:
that's right! Refactored

0xripleys commented on contracts/suilend/sources/obligation.move:
```diff
@@ -1221,4 +1370,57 @@ module suilend::obligation {
     }
-}
+
+    fun get_ltvs<P>(
```
nit: this function is a bit hard to read because of all the branches. in general, I prefer to return early in a function to avoid nesting if-elses, if that makes sense.

0xLendlord replied:
Makes sense. Refactored

0xLendlord committed:
- Fixed assertion over emode ltv thresholds
- Refresh after adding the emode flag
- Ability to toggle emode on and off
- Simplified refresh emode
- Fixed borrow weight when no emode pair
- Do not fail when getting ltvs
- Simplied get_ltvs

0xLendlord committed:
toggle emode tests

0xLendlord committed:
Nit: if-else statement

0xLendlord committed:
Add back toggle emode with reserve arrays
